engine/core/src/main/java/org/datacleaner/storage/InMemoryRowAnnotationFactory2.java

```diff
@@ -0,0 +1,155 @@
+ * along with this distribution; if not, write to:
+ * Free Software Foundation, Inc.
+ * 51 Franklin Street, Fifth Floor
+ * Boston, MA  02110-1301  USA
+ */
+package org.datacleaner.storage;
+
+import java.io.Serializable;
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.Collections;
+import java.util.List;
+import java.util.concurrent.ConcurrentHashMap;
+
+import org.datacleaner.api.InputRow;
+
+/**
+ * Successor of {@link InMemoryRowAnnotationFactory}, our implementation of
+ * {@link RowAnnotationFactory} that is based on in-memory storage of sample
+ * records. A new class was added to allow deserialization of old DataCleaner
+ * results, yet this class fully replaces the old one functionally.
+ */
+public final class InMemoryRowAnnotationFactory2 extends AbstractRowAnnotationFactory2 implements RowAnnotationFactory,
+        Serializable {
+
+    private static final long serialVersionUID = 1L;
+
+    /**
+     * Kind of a magic number, but a way to ensure that the ratio between sample
+     * sets and records is kept under control, at least in default scenarios.
+     */
+    private static final int DEFAULT_SAMPLE_LIMIT = 500 * 500;
+
+    private final ConcurrentHashMap<RowAnnotation, List<InputRow>> _storage;
+    private final int _maxSampleRecords;
+    private final int _maxSampleSets;
+
+    public InMemoryRowAnnotationFactory2() {
+        this(500);
+    }
+
+    public InMemoryRowAnnotationFactory2(int maxSampleRecords) {
+        this(Math.min(10, DEFAULT_SAMPLE_LIMIT / maxSampleRecords), 500);
+    }
+
+    /**
+     * 
+     * @param maxSampleSets
+     *            the maximum number of sample record collections to keep
+     * @param maxSampleRecords
+     *            the maximum number of records to keep in each collection
+     */
+    public InMemoryRowAnnotationFactory2(int maxSampleSets, int maxSampleRecords) {
+        _storage = new ConcurrentHashMap<>();
+        _maxSampleSets = Math.max(0, maxSampleSets);
+        _maxSampleRecords = Math.max(0, maxSampleRecords);
+    }
+
+    private void addInputRowsToCollection(Collection<InputRow> rowCollection, Collection<InputRow> rows) {
+        if (rowCollection == null) {
+            return;
+        }
+        int size = rowCollection.size();
+        if (size >= _maxSampleRecords) {
+            return;
+        }
+
+        for (InputRow inputRow : rows) {
+            rowCollection.add(inputRow);
+            size++;
+            if (size >= _maxSampleRecords) {
+                return;
+            }
+        }
+    }
+
+    private Collection<InputRow> getInputRowCollection(int defaultSize, RowAnnotation annotation) {
+        List<InputRow> rowCollection = _storage.get(annotation);
+        if (rowCollection == null) {
+            if (_storage.size() >= _maxSampleSets) {
+                return null;
+            }
+            rowCollection = new ArrayList<InputRow>(defaultSize);
```

Comment:
Also, ConcurrentHashMap has a setting for the maximum number of concurrent modifications, which you can tweak (or just put to one) to optimize memory usage as it is doing a concurrency/memory trade off.